From 62022814b3c4c0caa75971961eacdf53871148a6 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Tue, 27 Dec 2016 16:51:09 +0100
Subject: [PATCH 0688/1345] telephony: mvebu: tdmmc: simplify determining
 silicon revision

commit  3526e82fcbb165abfe48954d33476fc64337b84b from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

* Introduce new field in mv_phone_dev structure and assign value for
  AXP/A8k.
* Get rid of tdmmc_ip_ver_get function and SoC-ID's fillers.
* Shorten enum values' names in order to increase readability.

Change-Id: Ibee0c3d378ca34da5be7bdb13726d66e60ead800
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35152
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/telephony/mvebu_phone/mv_phone.h     |    6 +++--
 drivers/telephony/mvebu_phone/mv_phone_dev.c |    5 +++-
 drivers/telephony/mvebu_phone/tdmmc/tdmmc.c  |   35 ++++++++------------------
 drivers/telephony/mvebu_phone/tdmmc/tdmmc.h  |   11 +++-----
 4 files changed, 22 insertions(+), 35 deletions(-)

diff --git a/drivers/telephony/mvebu_phone/mv_phone.h b/drivers/telephony/mvebu_phone/mv_phone.h
index 6984843..81dd8cf 100644
--- a/drivers/telephony/mvebu_phone/mv_phone.h
+++ b/drivers/telephony/mvebu_phone/mv_phone.h
@@ -217,7 +217,6 @@ struct mv_phone_params {
 
 struct mv_phone_data {
 	u8 spi_mode;
-	u32 family_id;
 	enum mv_phone_frame_ts frame_ts;
 };
 
@@ -267,6 +266,9 @@ struct mv_phone_dev {
 #ifdef CONFIG_MV_TDM_EXT_STATS
 	u32 pcm_stop_fail;
 #endif
+
+	/* TDMMC silicon revision */
+	enum tdmmc_ip_version tdmmc_ip_ver;
 };
 
 /* This enumerator defines the Marvell Units ID */
@@ -323,7 +325,7 @@ int tdm2c_init(void __iomem *base, struct device *dev,
 
 /* TDMMC */
 int tdmmc_init(void __iomem *base, struct device *dev, struct mv_phone_params *tdm_params,
-	       struct mv_phone_data *hal_data);
+	       struct mv_phone_data *hal_data, enum tdmmc_ip_version tdmmc_ip_ver);
 int tdmmc_intr_low(struct mv_phone_intr_info *tdm_intr_info);
 
 #endif /* _MV_PHONE_H_ */
diff --git a/drivers/telephony/mvebu_phone/mv_phone_dev.c b/drivers/telephony/mvebu_phone/mv_phone_dev.c
index 5dea184..3a457ca 100644
--- a/drivers/telephony/mvebu_phone/mv_phone_dev.c
+++ b/drivers/telephony/mvebu_phone/mv_phone_dev.c
@@ -323,7 +323,8 @@ static int tdm_hw_init(struct mv_phone_params *tdm_params)
 
 		break;
 	case MV_TDM_UNIT_TDMMC:
-		ret = tdmmc_init(priv->tdm_base, priv->dev, tdm_params, &hal_data);
+		ret = tdmmc_init(priv->tdm_base, priv->dev, tdm_params,
+				 &hal_data, priv->tdmmc_ip_ver);
 
 		/* Issue SLIC reset */
 		ret |= tdmmc_reset_slic();
@@ -1045,6 +1046,7 @@ static int mvebu_phone_probe(struct platform_device *pdev)
 			goto err_clk;
 
 		priv->irq_count = 1;
+		priv->tdmmc_ip_ver = TDMMC_REV1;
 
 		tal_set_if(&tdmmc_if);
 	}
@@ -1054,6 +1056,7 @@ static int mvebu_phone_probe(struct platform_device *pdev)
 		tdmmc_set_a8k_windows(&pdev->dev, priv->tdm_base);
 
 		priv->irq_count = 3;
+		priv->tdmmc_ip_ver = TDMMC_REV1;
 
 		tal_set_if(&tdmmc_if);
 	}
diff --git a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
index d07f880..904d8b7 100644
--- a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
+++ b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
@@ -109,24 +109,10 @@
 static struct tdmmc_mcdma_tx_desc *mcdma_tx_desc_ptr[TOTAL_CHAINS];
 static dma_addr_t mcdma_rx_desc_phys[TOTAL_CHAINS], mcdma_tx_desc_phys[TOTAL_CHAINS];
 static struct tdmmc_dram_entry def_dpram_entry = { 0, 0, 0x1, 0x1, 0, 0, 0x1, 0, 0, 0, 0 };
-static u32 ctrl_family_id;
+static enum tdmmc_ip_version ip_ver;
 static struct device *pdev;
 static void __iomem *regs;
 
-static enum tdmmc_ip_version tdmmc_ip_ver_get(u32 ctrl_family_id)
-{
-	switch (ctrl_family_id) {
-	case MV_65XX_DEV_ID:
-		return MV_COMMUNIT_IP_VER_ORIGIN;
-	case MV_78XX0:
-	case MV_88F66X0:
-	case MV_88F67X0:
-		return MV_COMMUNIT_IP_VER_REVISE_1;
-	default:
-		return MV_COMMUNIT_IP_VER_REVISE_1;
-	}
-}
-
 static void tdmmc_desc_chain_build(void)
 {
 	u32 chan, index, buff_size;
@@ -194,7 +180,7 @@ static void tdmmc_mcdma_mcsc_start(void)
 	}
 
 	/* Set Rx/Tx periodical interrupts */
-	if (tdmmc_ip_ver_get(ctrl_family_id) == MV_COMMUNIT_IP_VER_ORIGIN)
+	if (ip_ver == TDMMC_REV0)
 		writel(CONFIG_VOICE_PERIODICAL_INT_CONTROL_WA,
 		       regs + VOICE_PERIODICAL_INT_CONTROL_REG);
 	else
@@ -372,7 +358,8 @@ void tdmmc_show(void)
 }
 
 int tdmmc_init(void __iomem *base, struct device *dev,
-	       struct mv_phone_params *tdm_params, struct mv_phone_data *hal_data)
+	       struct mv_phone_params *tdm_params, struct mv_phone_data *hal_data,
+	       enum tdmmc_ip_version tdmmc_ip_ver)
 {
 	u16 pcm_slot, index;
 	u32 buff_size, chan, total_rx_desc_size, total_tx_desc_size;
@@ -387,7 +374,7 @@ int tdmmc_init(void __iomem *base, struct device *dev,
 	total_channels = tdm_params->total_channels;
 	prev_rx_buff = 0;
 	next_tx_buff = 0;
-	ctrl_family_id = hal_data->family_id;
+	ip_ver = tdmmc_ip_ver;
 	pdev = dev;
 
 	/* Check parameters */
@@ -657,7 +644,7 @@ int tdmmc_init(void __iomem *base, struct device *dev,
 
 	/* Keep the software workaround to enable TEN while set Fsync for none-ALP chips */
 	/* Enable TDM */
-	if (tdmmc_ip_ver_get(ctrl_family_id) == MV_COMMUNIT_IP_VER_ORIGIN)
+	if (ip_ver == TDMMC_REV0)
 		mv_phone_set_bit(regs + FLEX_TDM_CONFIG_REG, TDM_TEN_MASK);
 
 	dev_dbg(pdev, "%s: Exit\n", __func__);
@@ -706,7 +693,7 @@ void tdmmc_release(void)
 		mv_phone_reset_bit(regs + MCSC_GLOBAL_CONFIG_REG, MCSC_GLOBAL_CONFIG_MAI_MASK);
 
 		/* Disable TDM */
-		if (tdmmc_ip_ver_get(ctrl_family_id) == MV_COMMUNIT_IP_VER_ORIGIN)
+		if (ip_ver == TDMMC_REV0)
 			mv_phone_reset_bit(regs + FLEX_TDM_CONFIG_REG, TDM_TEN_MASK);
 
 		/* Disable PCLK */
@@ -766,7 +753,7 @@ void tdmmc_pcm_start(void)
 		writel(CONFIG_COMM_UNIT_TOP_MASK, regs + COMM_UNIT_TOP_MASK_REG);
 
 		/* Enable TDM */
-		if (tdmmc_ip_ver_get(ctrl_family_id) == MV_COMMUNIT_IP_VER_REVISE_1)
+		if (ip_ver == TDMMC_REV1)
 			mv_phone_set_bit(regs + FLEX_TDM_CONFIG_REG, TDM_TEN_MASK);
 	}
 }
@@ -801,7 +788,7 @@ void tdmmc_pcm_stop(void)
 			memset(rx_buff_virt[index], 0, buff_size);
 
 		/* Disable TDM */
-		if (tdmmc_ip_ver_get(ctrl_family_id) == MV_COMMUNIT_IP_VER_REVISE_1)
+		if (ip_ver == TDMMC_REV1)
 			mv_phone_reset_bit(regs + FLEX_TDM_CONFIG_REG, TDM_TEN_MASK);
 	}
 }
@@ -814,7 +801,7 @@ int tdmmc_tx(u8 *tdm_tx_buff)
 	/* Calculate total Tx buffer size */
 	buff_size = (sample_size * MV_TDM_TOTAL_CH_SAMPLES * sampling_coeff * total_channels);
 
-	if (tdmmc_ip_ver_get(ctrl_family_id) == MV_COMMUNIT_IP_VER_ORIGIN) {
+	if (ip_ver == TDMMC_REV0) {
 		if (sample_size > MV_PCM_FORMAT_1BYTE) {
 			dev_dbg(pdev, "Linear mode (Tx): swapping bytes\n");
 			for (index = 0; index < buff_size; index += 2) {
@@ -837,7 +824,7 @@ int tdmmc_rx(u8 *tdm_rx_buff)
 	/* Calculate total Rx buffer size */
 	buff_size = (sample_size * MV_TDM_TOTAL_CH_SAMPLES * sampling_coeff * total_channels);
 
-	if (tdmmc_ip_ver_get(ctrl_family_id) == MV_COMMUNIT_IP_VER_ORIGIN) {
+	if (ip_ver == TDMMC_REV0) {
 		if (sample_size > MV_PCM_FORMAT_1BYTE) {
 			dev_dbg(pdev, "Linear mode (Rx): swapping bytes\n");
 			for (index = 0; index < buff_size; index += 2) {
diff --git a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h
index 1ed369a..3360711 100644
--- a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h
+++ b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h
@@ -343,12 +343,6 @@
 #define	CONFIG_TDM_DATA_DELAY_AND_CLK_CTRL \
 	(TX_CLK_OUT_ENABLE_MASK | RX_CLK_OUT_ENABLE_MASK)
 
-/* SoC ID's */
-#define MV_65XX_DEV_ID		0x6500	/* MV88F6500 family */
-#define MV_78XX0		0x78000	/* Armada XP Family */
-#define MV_88F66X0		0x6600	/* Avanta LP Family */
-#define MV_88F67X0		0x6700	/* Armada 375 Family */
-
 /* Defines */
 #define TOTAL_CHAINS		2
 #define CONFIG_RBSZ		16
@@ -359,9 +353,10 @@
 #define OLD_INT_WA_BIT		(1 << 15)
 #define MV_TDM_PCM_CLK_8MHZ	1
 
+/* Enums */
 enum tdmmc_ip_version {
-	MV_COMMUNIT_IP_VER_ORIGIN   = 0,
-	MV_COMMUNIT_IP_VER_REVISE_1,
+	TDMMC_REV0 = 0,
+	TDMMC_REV1
 };
 
 /* Structures */
-- 
1.7.9.5

